### PR TITLE
Add SCA prompt if required prior to "Change payment method" form submission

### PR DESCRIPTION
**Pull request by dechov**
# Changes proposed in this Pull Request:

Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046

Branches off of https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1398

The main change (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/38733a56779239e268cc7ef7f2a06b038cc342b8) is only a few lines of code, to expand the SCA prompt for the "Add payment method" form (in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1398) to the "Change payment method" form for subscriptions as well. (The preceding two commits are refactors to the relevant code that are not necessary for or directly related to the main fix – happy to drop them if that's better for any reason.)

This approach entirely supercedes https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1405, which redirects to an SCA step after saving the changed payment method – this is less desirable than authenticating prior to form submission. The reason I'd initially gone with approach in #1405 was to keep it similar to the "Pay for order" form, which seemed like a good idea at the time because it is so similar to the "Change payment method" form: the latter uses the same route, clears the former from the output buffer, and replaces it with a similar form (with the same ID).

However, with regard to SCA, it is more similar to the "Add payment method" form in terms of setting up the payment method outside the context of a payment (using a SetupIntent) and so there may not be any reason to follow the precedent of calling `process_payment` prior to the SCA challenge. Therefore, the exact same approach is borrowed, which addresses all of the remaining concerns in #1405. Assuming it's correct that there is no benefit to delaying the SCA prompt until after the form is submitted, the only caveat is just that this implementation has to look for a hidden input inside the form (to distinguish this form from the "Pay for order" form rather than checking the identifier (which is the kind of somewhat more brittle condition I'd originally hoped to avoid).

After attempting form submission with a card that requires SCA, the prompt appears over the payment form (here using [regulatory test card](https://stripe.com/docs/testing#regulatory-cards) in test mode):

After failure:

After success:

# Testing instructions

From #1405:

> Use instructions from https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046 but **verify that modal authentication appears**, and that if it succeeds then **the renewal also succeeds**. (Just note that the "Edit Subscription" screen needs to be reloaded after the payment method is saved, or else the "Process renewal" action will actually update it back to the earlier source, being the value of the "Stripe Source ID" field in the admin form 🤔.)

> (To change the payment method, navigate on the front-end to My account » Subscriptions » any active subscription and click "Change payment", or navigate in wp-admin to WooCommerce » Subscriptions » any active subscription and click "Customer change payment method page →".)

> Can also test to make sure the ordinary Pay for Order flow (for regular pending orders) still works as before.

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: assets/js/stripe.js [hunk 1]
```diff
@@ -531,7 +531,7 @@ jQuery( function( $ ) {
 				return;
 			}
 
-			wc_stripe_form.form.submit();
```
Comment (DanReyLop) [inline assets/js/stripe.js]: Nice cleanup! I can't even find it's deprecated in jQuery's documentation, but I remember reading something about it :)

Comment (dechov) [reply assets/js/stripe.js]: Forgot to mention this was changed in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1390 but then (seemingly unintentionally) changed back in #1398.

**Comment by DanReyLop** [comment]
> The preceding two commits are refactors to the relevant code that are not necessary for or directly related to the main fix – happy to drop them if that's better for any reason.

The PR has 15 added lines total. The refactors can stay :)

> Assuming it's correct that there is no benefit to delaying the SCA prompt until after the form is submitted

I agree with that assumption, I don't see any reason to delay the prompt.

> The only caveat is just that this implementation has to look for a hidden input inside the form (to distinguish this form from the "Pay for order" form rather than checking the identifier (which is the kind of somewhat more brittle condition I'd originally hoped to avoid).

I get that worry. Also, this isn't a WC Core form, but a WC Subscriptions form, so we depend on the HTML stability of WC Subscriptions.

I suppose you could hook into `woocommerce_subscriptions_change_payment_before_submit` and render a dummy hidden `<input>` that can act as a "beacon", and because that filter is part of WC Subscription's "public API", it's much less likely to break. I can see it being considered overkill, but I wouldn't be opposed if you implemented it :)

Code looks good, I'll test the changes now.

**Comment by dechov** [comment]
> I suppose you could hook into woocommerce_subscriptions_change_payment_before_submit and render a dummy hidden <input> that can act as a "beacon", and because that filter is part of WC Subscription's "public API", it's much less likely to break.

Thanks @DanReyLop! I've done just that in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1427/commits/83fa93cf75f1d3bedfdd4b026455b4881a4ce690

**Review by DanReyLop — APPROVED** [review]
Code looks good and it works as expected. :shipit: !

**Pull request merged**